packages/app-desktop/gui/MenuBar.tsx

```diff
@@ -320,12 +316,13 @@ function useMenu(props: Props) {
 			format: module.format,
 			outputFormat: module.outputFormat,
 			// eslint-disable-next-line @typescript-eslint/no-explicit-any -- Old code before rule was applied
-			onProgress: (status: any) => {
-				const statusStrings: string[] = Object.keys(status).map((key: string) => {
-					return `${key}: ${status[key]}`;
+			onProgress: (_status: any) => {
+				props.dispatch({
+					type: 'INTEROP_EXEC',
```

Comment:
```suggestion
					type: 'UPDATE_INTEROP_TASK_STATUS',
```

Consider using a different name here.

Reply:
Should I also create an action for this in app.reducer? The onProgress property is currently not used, so that action is never dispatched. We just left it there as you could be working on something that would use that function.

Reply:
> Also, I think there is no way to dynamically update the notifications. The only way I can think of is by dismissing and opening a new notification every time we receive a new interop update through state mutation.

One option (maybe not a good one :) ) could be to avoid using the reducer altogether and instead create a new `TaskProgressUIService` in [lib/services/](https://github.com/laurent22/joplin/tree/dev/packages/lib/services).

`TaskProgressUIService` could look similar to this:
```ts
// TaskProgressUIService might not be the best name for this...
class TaskProgressUIService {
  public addTaskStartListener(listener: ...): ...;
  public addTaskProgressListener(...): ...;
  public addTaskCompleteListener(...): ...;

  public onTaskStarted(id: string, name: string): void;
  public onTaskProgressed(id: string, progress: number): void;
  public onTaskCompleted(id: string): void;
}
```

A hook could be created (in [packages/lib/hooks](https://github.com/laurent22/joplin/tree/dev/packages/lib/hooks)?) to use the current running tasks and their progress. It could be implemented by adding and removing listeners from `TaskProgressUIService`.

The notification component could then use this hook to determine the content of the notification(s) to be shown.

**Reasons this approach could be better**:
- **Notifications for other tasks**: `TaskProgressUIService` could also be used to resolve https://github.com/laurent22/joplin/issues/10527 (though in a separate pull request).
- **Progress notifications on mobile**:
    - `app-mobile` could also use `TaskProgressUIService` to show notifications. Note that mobile uses the `InteropService` for import/export of JEX files, but shows progress/task status differently from desktop.
- **Avoids adding additional action handlers to the reducer**: (This might just be personal opinion.)
    - Currently, reducer-related code isn't as type safe as it could be (it's easy to miss-type reducer action keys).
    - Relying on global reducer actions can make it more difficult to write tests for components, as the reducer logic needs to be ported to a testing environment. While I think this has been done for the reducer logic in `lib/`, I don't think this has been done for the desktop or mobile reducers.

Reply:
What I meant to say was that notyf in its current state cannot dynamically update the content of the notification: the only way we could implement progress using the notyf API is by dismissing the notification and opening another one with the updated progress percentage, which may not be what we want.
However I understand the idea behind creating some kind of a TaskProgressUIService and maybe we could try to do something like that if we find a way of creating progress notifications.

Reply:
One option could be to do this in a way similar to how [the "deleted to trash" notification adds a clickable link to its notification](https://github.com/laurent22/joplin/blob/97ddb67f683cc4ed3ad832c2f97ce0e8a3057528/packages/app-desktop/gui/TrashNotification/TrashNotification.tsx#L73) (custom HTML content with an ID and `@joplin/lib/dom`'s [`waitForElement`](https://github.com/laurent22/joplin/blob/dev/packages/lib/dom.ts#L11)).

Alternatively, I wonder if it would make sense to create a custom notification component — live updates seem to be [outside the scope of `notyf`](https://github.com/caroso1222/notyf/issues/73#issuecomment-655090322). A drawback, however, would be that these notifications might cover (or be covered by) the existing trash notifications.

<!-- Additionally, `notyf` may not be maintained — its [last commit was three years ago](https://github.com/caroso1222/notyf/commit/d19f0abf35d3b785e8fec6c81a08e18087feac56), though its maintainer has [commented on issues in the repository more recently](https://github.com/caroso1222/notyf/issues?q=commenter%3Acaroso1222+sort%3Aupdated-desc). -->

Thank you for clarifying!

Reply:
Hi, we implemented the TaskProgressUIService and a TaskNotification that can be updated during the designed process. But we are encountering one problem:

So during the import, for example, we need to call onTaskProgress that updates the task's progress and the notification related to that task:

```
public onTaskProgress(id: string, progress: number): void {
		const task = this.tasks.get(id);
		if (task.progress !== 100) {
			task.progress = progress % 100;
		}
		if (this.listener) {
			this.listener(task);
		} else {
			this.queue.add(id);
		}
	}
```

The import is processed by a InteropService that is called with importOptions (these options contain the onProgress function: onTaskProgress).
```
onProgress: (_status: any) => { TaskProgressUIService.onTaskProgress(taskId, _status.percentage); },
```
Then the interopService instances a importer:
```
const importer = this.newModuleByFormat_(ModuleType.Importer, options.format, options.outputFormat); 
```
That does the process:
```
await importer.init(options.path, options);
result = await importer.exec(result);
```

Our problem is that we can't get status from this importer.exec, which doesn't allow us to create a dynamic progress bar/circle. We could use stages to do it, but would only contain (init (0%), started (5%), and then finish (100%)) which would not have a good screen effect.

One option that we think could be a great solution for this is using a circle animation  that would still have the effect of an ongoing process. Just wouldn't have the percentage as we would like to have.

What do you think of this solution? Do you have any sugestions?

Reply:
> One option that we think could be a great solution for this is using a circle animation that would still have the effect of an ongoing process. Just wouldn't have the percentage as we would like to have.

Indeed, only some interop tasks currently have progress information available. One option could be to continue using an hourglass icon when no progress information is available (i.e. show no progress indicator).

Another option could be to show an indeterminate progress indicator until progress is set for the first time.
<details><summary>Example CSS</summary>

HTML:
```html
<div class='loading-spinner -indeterminate'></div>
```

Note: The first call to `onTaskProgress` could remove the `-indeterminate` class from `loading-spinner`.

CSS:
```css
@keyframes spin {
    from { transform: rotate(360deg); }
    to { transform: rotate(0deg); }
}
.loading-spinner {
    display: inline-block;
    width: 15px;
    height: 15px;

    border: 3px solid var(--joplin-background-color, white);
    border-bottom-color: var(--joplin-color, black);
    border-radius: 15px;
}
.loading-spinner.-indeterminate {
    animation: 4s linear infinite spin;
}
@media (prefers-reduced-motion) {
    .loading-spinner.-indeterminate { animation: none; }
}
```

See also:
- https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion
- https://developer.mozilla.org/en-US/docs/Web/CSS/animation

</details>